### PR TITLE
Implement sketched percentile

### PR DESCRIPTION
**Pull request by mrfh92**
This solves #1411 by implementing a sketched version of percentile / median and a corresponding option for the RobustScaler in the preprocessing module.   

All new features integrate into the current functionality without changes, due to choice of the current non-sketched version as default. 

**Performance comparison**
on 14 MPI-processes (Workstation, CPU) on behalf of the "asteroids data set" (1317275 data points, 9 features), times and deviation measured over10 runs 

| | avg time | inaccuracy* |
|--------|--------|--------|
| median main | ~12.3s | |
| median sketched (10% of data) | 0.247s | 0.30% | 
| median sketched (1% of data) | 0.01212s | 1.3% |
| median sketched (0.1% of data) | 0.00312s | 2.7% |

*relative deviation of the sketched median vs true median per feature, averaged over 10 runs; only relative deviation for the feature with maximal deviation is reported

### CODE REVIEW

File: heat/core/tests/test_statistics.py [hunk 1]
```diff
@@ -1244,6 +1244,28 @@ def test_percentile(self):
...
+                self.assertTrue(p.shape == (2 * ht.MPI_WORLD.size,))
+        # check if it raises correct errors
+        X = ht.random.rand(10 * ht.MPI_WORLD.size, 2 * ht.MPI_WORLD.size, split=0)
+        axis = 0
+        q = 50
+        with self.assertRaises(ValueError):
+            ht.percentile(X, q, axis=axis, use_sketch_of_size=1.1)
+        with self.assertRaises(ValueError):
+            ht.percentile(X, q, axis=axis, use_sketch_of_size=10)
+
```
Comment (mtar) [inline heat/core/tests/test_statistics.py:1268]:
```suggestion
    def test_percentile_sketched(self):
        axis, q = 0, 50
        use_sketch_of_size = 0.1
        q = 50
        # check if it works
        for split in [None, 1, 0]:
            X = ht.random.rand(10 * ht.MPI_WORLD.size, 2 * ht.MPI_WORLD.size, split=split)
            if axis != split:
                with self.assertWarns(Warning):
                    ht.percentile(X, q, axis=axis, use_sketch_of_size=use_sketch_of_size)
            else:
                p = ht.percentile(X, q, axis=axis, use_sketch_of_size=use_sketch_of_size)
                self.assertTrue(p.shape == (2 * ht.MPI_WORLD.size,))
        # check if it raises correct errors
        with self.assertRaises(ValueError):
            ht.percentile(X, q, axis=axis, use_sketch_of_size=1.1)
        with self.assertRaises(ValueError):
            ht.percentile(X, q, axis=axis, use_sketch_of_size=10)

```

**Review by mtar — COMMENTED** [review]
Some minor comments on documentation

File: heat/core/statistics.py [hunk 2]
```diff
@@ -1032,12 +1039,17 @@ def median(x: DNDarray, axis: Optional[int] = None, keepdims: bool = False) -> D
     keepdims : bool, optional
         If True, the axes which are reduced are left in the result as dimensions with size one.
         With this option, the result can broadcast correctly against the original array ``a``.
+
+    use_sketch_of_size : float, optional
+        The fraction of the data to use for estimating the median. If None, the entire data is used.
```
[the remark this answers is not in the document]
Comment (mrfh92) [reply heat/core/statistics.py]: done

File: heat/core/statistics.py [hunk 3]
```diff
@@ -1447,6 +1462,9 @@ def percentile(
     keepdims : bool, optional
         If True, the axes which are reduced are left in the result as dimensions with size one.
         With this option, the result can broadcast correctly against the original array x.
+
+    use_sketch_of_size : float, optional
+        The fraction of the data to use for estimating the percentiles. If None, the entire data is used.
```
[the remark this answers is not in the document]
Comment (mrfh92) [reply heat/core/statistics.py]: done

File: heat/core/statistics.py [hunk 4]
```diff
@@ -1412,10 +1424,13 @@ def percentile(
     out: Optional[DNDarray] = None,
     interpolation: str = "linear",
     keepdims: bool = False,
+    use_sketch_of_size: Optional[float] = None,
 ) -> DNDarray:
     r"""
     Compute the q-th percentile of the data along the specified axis.
     Returns the q-th percentile(s) of the tensor elements.
+    If use_sketch_of_size is not None, only a random subset of the data is used to estimate the percentiles which
+    may result in inaccurate results but is faster.
```
[the remark this answers is not in the document]
Comment (mrfh92) [reply heat/core/statistics.py]: made the explanation a bit more precise; range for the values is given in the parameters section of the docstring.

File: heat/core/statistics.py [hunk 5]
```diff
@@ -1016,10 +1018,18 @@ def reduce_means_elementwise(output_shape_i: torch.Tensor) -> DNDarray:
 DNDarray.mean.__doc__ = mean.__doc__
 
 
-def median(x: DNDarray, axis: Optional[int] = None, keepdims: bool = False) -> DNDarray:
+def median(
+    x: DNDarray,
+    axis: Optional[int] = None,
+    keepdims: bool = False,
+    use_sketch_of_size: Optional[float] = None,
```
Comment (ClaudiaComito) [inline heat/core/statistics.py:1025]: Could we use two kwargs here for clarity? For example
```python
sketched: bool = False,
sketch_size: Optional[float] = 1.,
```

(same for percentile and preprocessing)

File: heat/core/statistics.py [hunk 6]
```diff
@@ -1490,13 +1514,72 @@ def _local_percentile(data: torch.Tensor, axis: int, indices: torch.Tensor) -> t
...
+        # if isinstance(sketch_size, int) and sketch_size <= 0 or sketch_size > a.shape[axis]:
+        #     raise ValueError(
+        #         f"if an integer, sketch_size must be positive and less than or equal to a.shape[axis], but is {sketch_size}."
+        #     )
+        if isinstance(sketch_size, float):
+            # if sketch_size <= 0 or sketch_size >= 1:
+            #     raise ValueError(
+            #         f"if a float, sketch_size must be strictly between 0 and 1, but is {sketch_size}."
+            #     )
+            # else:
```
Comment (ClaudiaComito) [inline heat/core/statistics.py:1547]: dead code?

File: heat/core/statistics.py [hunk 7]
```diff
@@ -1490,13 +1514,72 @@ def _local_percentile(data: torch.Tensor, axis: int, indices: torch.Tensor) -> t
...
+            sketch_size = int(sketch_size * a.shape[axis])
+        # if not isinstance(sketch_size, int) and not isinstance(sketch_size, float):
+        #     raise ValueError(
+        #         f"sketch_size must be an integer or a float, but is {type(sketch_size)}."
+        #     )
+
```
Comment (ClaudiaComito) [inline heat/core/statistics.py:1553]: dead code?

**Review by ClaudiaComito — CHANGES_REQUESTED** [review]
This is a great feature to have, thanks a lot @mrfh92 ! A few comments below...

File: heat/core/statistics.py [hunk 8]
```diff
@@ -1490,13 +1514,72 @@ def _local_percentile(data: torch.Tensor, axis: int, indices: torch.Tensor) -> t
...
+        # create a random sample of indices
+        indices = randint(0, a.shape[axis], sketch_size, device=a.device)
```
[the remark this answers is not in the document]
Comment (mrfh92) [reply heat/core/statistics.py:1555]: I suggest to keep this as it is as this problem will be addressed in the refactoring of the random-module anyway

File: heat/core/statistics.py (hunk 6 above)
Comment (mrfh92) [reply heat/core/statistics.py:1547]: removed

File: heat/core/statistics.py (hunk 7 above)
Comment (mrfh92) [reply heat/core/statistics.py:1553]: removed

File: heat/core/statistics.py [hunk 9]
```diff
@@ -1490,13 +1514,72 @@ def _local_percentile(data: torch.Tensor, axis: int, indices: torch.Tensor) -> t
...
+        Create a sketch of a DNDarray along a specified axis. The sketch is created by sampling the DNDarray along the specified axis.
+
+        Parameters
+        ----------
+        a : DNDarray
+            The DNDarray for which to create a sketch.
+        axis : int
+            The axis along which to create the sketch.
+        sketch_size : int or float
+            The size of the sketch. If an integer, it is interpreted as the number of samples to take and therefore must not exceed the size of the axis along which the sketch is taken. If a float, it is interpreted as the fraction of samples to take and must be strictly between 0 and 1.
```
[the remark this answers is not in the document]
Comment (mrfh92) [reply heat/core/statistics.py:1530]: I have chosen the mutually exclusive two arguments.

File: heat/core/statistics.py [hunk 10]
```diff
@@ -1490,13 +1514,72 @@ def _local_percentile(data: torch.Tensor, axis: int, indices: torch.Tensor) -> t
...
     if not isinstance(x, DNDarray):
         raise TypeError(f"expected x to be a DNDarray, but was {type(x)}")
     if isinstance(axis, (list, tuple)):
         raise NotImplementedError("ht.percentile(), tuple axis not implemented yet")
 
+    if use_sketch_of_size is not None:
+        if axis != x.split:
+            warn(
+                f"Sketching is only advantageous in the case that axis, i.e. the axis along which to take the percentile, and the split axis of the array coincide. Consider avoiding the sketching option in the present since axis={axis}, x.split={x.split}."
+            )
```
[the remark this answers is not in the document]
Comment (mrfh92) [reply heat/core/statistics.py:1571]: removed the warning

File: heat/core/statistics.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ClaudiaComito) [reply heat/core/statistics.py:1549]: :+1:

File: heat/core/statistics.py (hunk 5 above)
Comment (mrfh92) [reply heat/core/statistics.py:1025]: taken into account, but with sketch_size = 1./MPI.COMM_WORLD.size as default as discussed this morning

File: heat/core/statistics.py [hunk 11]
```diff
@@ -1016,10 +1018,19 @@ def reduce_means_elementwise(output_shape_i: torch.Tensor) -> DNDarray:
...
+    axis: Optional[int] = None,
+    keepdims: bool = False,
+    sketched: bool = False,
+    sketch_size: Optional[float] = 1.0 / MPI.COMM_WORLD.size,
+) -> DNDarray:
     """
     Compute the median of the data along the specified axis.
     Returns the median of the ``DNDarray`` elements.
+    Per default, the "true" median of the entire data set is computed; however, the argument
+    `sketched` allows to switch to a faster but inaccurate version that computes
```
[the remark this answers is not in the document]
Comment (ClaudiaComito) [reply heat/core/statistics.py:1032]: How about showing your accuracy measurements from the PR description in a Notes paragraph? We can say that it's just an example and results will vary with dataset, but just as an example. Or do you think it's not representative enough?

**Review by ClaudiaComito — CHANGES_REQUESTED** [review]
Just a couple tiny changes, otherwise I think we can merge this week. Thanks a lot @mrfh92 !

**Comment by mrfh92** [comment]
I think showing the example accuracies does not make so much sense since these accuracies very much depend on the data set (even if the median is quite stable w.r.t. outliers) and usually, we only show examples for small data that would allow calculation by hand too

**Review by ClaudiaComito — APPROVED** [review]
Great, thanks @mrfh92 !

**Comment by mrfh92** [comment]
@ClaudiaComito I have tried to fix the bug on torch 1.12 and 1.13

**Review by ClaudiaComito — APPROVED** [review]

**Pull request merged**